lib/createDuck.ts

```diff
@@ -1,36 +1,40 @@
-export type Action<TPayload> = {
+export interface BaseAction {
     type: string;
-    payload?: TPayload;
-};
+}
 
-export type PayloadReducer<TState, TPayload> = {
-    (state: TState, payload: TPayload): TState;
-};
+export interface Action<TPayload> extends BaseAction {
+    type: string;
+    payload: TPayload;
+}
```

Comment:
I suggest to not use inheritance here. Even though, you are overwriting the `type` property already. 😉

And what is the benefit of having `interface` instead of `type` definitions here? See also: https://stackoverflow.com/questions/37233735/typescript-interfaces-vs-types

Reply:
I used interfaces because I needed the extending to split the `Action` type into one with and without payload.
The reason is that when you use strict null checking (which I do in my project), it is really annoying to always tell the compiler with additional code that you already know the payload is not undefined.
This change properly splits actions with and without payloads.
Only after writing this I noticed that at least writing something like `action.payload!.myData`, however I don't think you could call that a clean solution.